### PR TITLE
Portable stories: Make setProjectAnnotations accept multiple types of imports

### PR DESCRIPTION
**Pull request by yannbf**
Closes #

## What I did

Before, it was tricky for users to know how to set project annotations because they can either come from a file such as:

1 -
```ts
// .storybook/preview.ts
const preview = {
	parameters: {}
}

export default preview;
```

2 -
```ts
// .storybook/preview.ts
export const parameters = {}
```

Meaning they would have to know whether to import them like so (in a Vitest setup file for instance):
```ts
// setup-file.ts
import * as annotations from '../.storybook/preview' // or some-addon/preview
import { setProjectAnnotations } from '@storybook/react'
setProjectAnnotations(annotations)
```

or like so:
```ts
// setup-file.ts
import annotations from '../.storybook/preview' // or some-addon/preview
import { setProjectAnnotations } from '@storybook/react'
setProjectAnnotations(annotations)
```

It gets tricky because annotations can come differently from where users are importing.

This PR fixes that issue by accepting both formats, and in the documentation we will be telling users to use `import * as` even if the file has just a default export. This way, the code will always be concise such as:

```ts
import * as sbAnnotations from '../.storybook/preview'
import * as addonAnnotations from 'some-addon/preview'
import { setProjectAnnotations } from '@storybook/react'
setProjectAnnotations([sbAnnotations, addonAnnotations])
```

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

### CODE REVIEW

File: code/lib/preview-api/src/modules/store/csf/portable-stories.ts [hunk 1]
```diff
@@ -24,23 +25,34 @@ import { normalizeComponentAnnotations } from './normalizeComponentAnnotations';
...
 export function getPortableStoryWrapperId(storyId: string) {
   return `storybook-story-${storyId}`;
 }
 
+function extractAnnotation<TRenderer extends Renderer = Renderer>(
+  annotation: FlexibleProjectAnnotations<TRenderer>
+) {
+  // support imports such as
+  // import * as annotations from '.storybook/preview'
+  // in both cases: 1 - the file has a default export; 2 - named exports only
```
Comment (JReinhold) [inline code/lib/preview-api/src/modules/store/csf/portable-stories.ts:39]: nit: feel like there's missing a line here. Referring to "both cases" but only one case is described.

File: code/lib/types/src/modules/story.ts [hunk 2]
```diff
@@ -47,6 +47,11 @@ export type ProjectAnnotations<TRenderer extends Renderer> = CsfProjectAnnotatio
   renderToDOM?: RenderToCanvas<TRenderer>;
 };
 
+// TODO: find a better name for this, of course. Please help
```
Comment (JReinhold) [inline code/lib/types/src/modules/story.ts:50]: I don't think a good name exists for this type of thing. `NamedOrDefaultProjectAnnotations` perhaps?

In _reality_ this kind of type would explain it well I think:

```
export type NamedExportsOrDefault<TExport> =
  | TExport
  | { default: TExport };

...
// somewhere else:

type ProjectAnnotations = NamedExportsOrDefault<OriginalProjectAnnotations<TRenderer>>;
```

But that might be over-engineering it.

**Review by JReinhold — APPROVED** [review]
🚀

File: code/lib/preview-api/src/modules/store/csf/portable-stories.test.ts [hunk 3]
```diff
@@ -23,6 +23,37 @@ describe('composeStory', () => {
...
+      {
+        // import * as annotations from '.storybook/preview'
+        default: {
+          parameters: {
+            fromAnnotations: {
+              asDefaultImport: true,
+            },
+          },
+        },
+      },
```
Comment (ndelangen) [inline code/lib/preview-api/src/modules/store/csf/portable-stories.test.ts:45]: Let's extract these into 2 files actually imported into the test file.

```
code/lib/preview-api/src/modules/store/csf/portable-stories.mock-a.ts
code/lib/preview-api/src/modules/store/csf/portable-stories.mock-b.ts
```

File: code/lib/types/src/modules/composedStory.ts [hunk 4]
```diff
@@ -76,7 +76,7 @@ export interface ComposeStoryFn<TRenderer extends Renderer = Renderer, TArgs ext
   (
     storyAnnotations: AnnotatedStoryFn<TRenderer, TArgs> | StoryAnnotations<TRenderer, TArgs>,
     componentAnnotations: ComponentAnnotations<TRenderer, TArgs>,
-    projectAnnotations: ProjectAnnotations<TRenderer>,
+    projectAnnotations: FlexibleProjectAnnotations<TRenderer>,
```
Comment (ndelangen) [inline code/lib/types/src/modules/composedStory.ts:79]: I propose you do not "leak" the `NamedOrDefault` part down so far.

Let's convert `NamedOrDefaultProjectAnnotations` to `ProjectAnnotations` as soon as we can, and not let the underlaying code worry about this.

**Pull request merged**